flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionMemory.java

```diff
@@ -0,0 +1,130 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+
+import javax.annotation.Nullable;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Putting and getting of buffers to/from a region of memory.
+ * This class handles the headers, length encoding, memory slicing.
+ */
+final class BoundedBlockingSubpartitionMemory {
+
+	// all fields and methods below here have package-private access to avoid bridge
+	// methods when accessing them from the nested classes
+
+	static final int HEADER_LENGTH = 8;
+
+	static final int HEADER_VALUE_IS_BUFFER = 0;
+
+	static final int HEADER_VALUE_IS_EVENT = 1;
+
+	static ByteBuffer checkAndConfigureByteBuffer(ByteBuffer buffer) {
+		checkArgument(buffer.position() == 0);
+		checkArgument(buffer.capacity() > 8);
+		checkArgument(buffer.limit() == buffer.capacity());
+
+		return buffer.order(ByteOrder.nativeOrder());
+	}
+
+	// ------------------------------------------------------------------------
+
+	static final class Writer {
+
+		private final ByteBuffer memory;
+
+		Writer(ByteBuffer memory) {
+			this.memory = checkAndConfigureByteBuffer(memory);
+		}
+
+		public boolean writeBuffer(Buffer buffer) {
+			final ByteBuffer memory = this.memory;
+			final int bufferSize = buffer.getSize();
+
+			if (memory.remaining() < bufferSize + HEADER_LENGTH) {
+				return false;
+			}
+
+			memory.putInt(buffer.isBuffer() ? HEADER_VALUE_IS_BUFFER : HEADER_VALUE_IS_EVENT);
+			memory.putInt(bufferSize);
+			memory.put(buffer.getNioBufferReadable());
+			return true;
+		}
+
+		public ByteBuffer complete() {
+			memory.flip();
+			return memory;
+		}
+
+		public int getNumBytes() {
+			return memory.position();
+		}
+	}
+
+	static final class Reader {
+
+		private final ByteBuffer memory;
+
+		Reader(ByteBuffer memory) {
+			this.memory = checkAndConfigureByteBuffer(memory);
+		}
+
+		@Nullable
+		public Buffer sliceNextBuffer() {
+			final ByteBuffer memory = this.memory;
+			final int remaining = memory.remaining();
+
+			// we only check the correct case where data is exhausted
+			// all other cases can only occur if our write logic is wrong and will already throw
+			// buffer underflow exceptions which will cause the read to fail.
+			if (remaining == 0) {
+				return null;
+			}
+
+			final int header = memory.getInt();
+			final int size = memory.getInt();
+
+			memory.limit(memory.position() + size);
+			ByteBuffer buf = memory.slice();
+			memory.position(memory.limit());
+			memory.limit(memory.capacity());
```

Comment:
here, is it necessary to `memory.limit(memory.capacity())` ?

Reply:
yes, it is, the code will not work without

Reply:
If we don't set the limit forward, the successive `readInt()` calls for the header will fail.
I played around with some alternatives, but could not find any where the code overall would become simpler.